java.lang.IllegalArgumentException: register too many Broadcast Receivers

<!--
Please fill in the below fields with some data to help us best diagnose the issue.
The more specific you are, the better! You can help a lot by not making us ask these questions.
Feel free to remove any irrelevant parts that you know are not related to the issue.
Any HTML comment like this will be stripped when rendering markdown, no need to delete them.
-->

<!-- What version of Glide you're running, for example: 3.7.1 | 3.8.0-SNAPSHOT | 4.0.0-SNAPSHOT
It's essentially the version number from your build.gradle: `dependencies { compile '...:x.y.z' }` -->

**Glide Version**: 3.6.1

<!-- What devices you managed to get the issue to come up on? For example:
fails on Galaxy S4/GT-I9500 4.4.2, works fine on Nexus 6P 5.1 and Genymotion Nexus 5 5.0.1 -->

**Device/Android Version**: Huawei RIO-L01 Android 5.1

<!-- Share the details of your issue in prose, detailing actual and expected behavior. It also helps if you give some info **why** you are trying to do something as opposed to **what** is not working. -->

**Issue details**: It seems to be caused by an issue in the Android framework that occurs only on Huawei RIO-L01. Anyway it could be solved catching the exception when `DefaultConnectivityMonitor` registers the receiver.

Actually I was considering to fix it myself providing a no-op `ConnectivityMonitor` only for the devices affected by this issue through `GlideBuilder#setConnectivityMonitorFactory`, but I have seen that this isn't possible in 3.x version.

<!--
What is the error message that you got in the log?
You can find some help on diagnosing issues here: https://github.com/bumptech/glide/wiki/Debugging-and-Error-Handling
-->

**Stack trace / LogCat**:

``` ruby
Fatal Exception: java.lang.RuntimeException: Unable to start activity ComponentInfo{it.subito/it.subito.v2.detail.AdDetailActivity}: java.lang.IllegalArgumentException: regist too many Broadcast Receivers
       at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2450)
       at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2512)
       at android.app.ActivityThread.access$1200(ActivityThread.java:165)
       at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1373)
       at android.os.Handler.dispatchMessage(Handler.java:102)
       at android.os.Looper.loop(Looper.java:135)
       at android.app.ActivityThread.main(ActivityThread.java:5593)
       at java.lang.reflect.Method.invoke(Method.java)
       at java.lang.reflect.Method.invoke(Method.java:372)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:960)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:755)
Caused by java.lang.IllegalArgumentException: regist too many Broadcast Receivers
       at android.app.ContextImpl.registerReceiverInternal(ContextImpl.java:1836)
       at android.app.ContextImpl.registerReceiver(ContextImpl.java:1794)
       at android.app.ContextImpl.registerReceiver(ContextImpl.java:1788)
       at android.content.ContextWrapper.registerReceiver(ContextWrapper.java:489)
       at com.bumptech.glide.manager.DefaultConnectivityMonitor.register(SourceFile:39)
       at com.bumptech.glide.manager.DefaultConnectivityMonitor.onStart(SourceFile:61)
       at com.bumptech.glide.manager.ActivityFragmentLifecycle.onStart(SourceFile:50)
       at com.bumptech.glide.manager.SupportRequestManagerFragment.onStart(SourceFile:135)
       at android.support.v4.app.Fragment.performStart(SourceFile:1986)
       at android.support.v4.app.FragmentManagerImpl.moveToState(SourceFile:1102)
       at android.support.v4.app.FragmentManagerImpl.moveToState(SourceFile:1248)
       at android.support.v4.app.FragmentManagerImpl.moveToState(SourceFile:1230)
       at android.support.v4.app.FragmentManagerImpl.dispatchStart(SourceFile:2047)
       at android.support.v4.app.FragmentController.dispatchStart(SourceFile:176)
       at android.support.v4.app.FragmentActivity.onStart(SourceFile:553)
       at com.trello.rxlifecycle.components.support.RxAppCompatActivity.onStart(SourceFile:43)
       at it.subito.v2.activities.BaseActivity.onStart(SourceFile:41)
       at android.app.Instrumentation.callActivityOnStart(Instrumentation.java:1236)
       at android.app.Activity.performStart(Activity.java:6118)
       at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2413)
       at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2512)
       at android.app.ActivityThread.access$1200(ActivityThread.java:165)
       at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1373)
       at android.os.Handler.dispatchMessage(Handler.java:102)
       at android.os.Looper.loop(Looper.java:135)
       at android.app.ActivityThread.main(ActivityThread.java:5593)
       at java.lang.reflect.Method.invoke(Method.java)
       at java.lang.reflect.Method.invoke(Method.java:372)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:960)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:755)
```

<!-- Bonus points if you attach a relevant screenshot, screen recording or a small demo project -->

We are also facing same Issue of glide.

**Glide Version** : 4.11
**Device :** LENOVO Lenovo TB-X605FC

**Description** : We are using BaseActivity to load image and using base activity context. here is code snippet...

Stack Trace Log Cat:

```
java.lang.RuntimeException: Unable to start activity ComponentInfo{com.grubbrrsystems.kiosk/com.grubbrrsystems.kiosk.general.NormalItemDetailsActivity}: java.lang.IllegalStateException: Too many receivers, total of 1000, registered for pid: 21747, callerPackage: com.grubbrrsystems.kiosk
	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3280)
	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3419)
	at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:83)
	at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135)
	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2020)
	at android.os.Handler.dispatchMessage(Handler.java:107)
	at android.os.Looper.loop(Looper.java:214)
	at android.app.ActivityThread.main(ActivityThread.java:7406)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:944)
Caused by: java.lang.IllegalStateException: Too many receivers, total of 1000, registered for pid: 21747, callerPackage: com.grubbrrsystems.kiosk
	at android.os.Parcel.createException(Parcel.java:2079)
	at android.os.Parcel.readException(Parcel.java:2039)
	at android.os.Parcel.readException(Parcel.java:1987)
	at android.app.IActivityManager$Stub$Proxy.registerReceiver(IActivityManager.java:4824)
	at android.app.ContextImpl.registerReceiverInternal(ContextImpl.java:1527)
	at android.app.ContextImpl.registerReceiver(ContextImpl.java:1488)
	at android.app.ContextImpl.registerReceiver(ContextImpl.java:1476)
	at android.content.ContextWrapper.registerReceiver(ContextWrapper.java:627)
	at com.bumptech.glide.manager.DefaultConnectivityMonitor.register(DefaultConnectivityMonitor.java:60)
	at com.bumptech.glide.manager.DefaultConnectivityMonitor.onStart(DefaultConnectivityMonitor.java:106)
	at com.bumptech.glide.manager.ActivityFragmentLifecycle.addListener(ActivityFragmentLifecycle.java:37)
	at com.bumptech.glide.RequestManager.<init>(RequestManager.java:144)
	at com.bumptech.glide.RequestManager.<init>(RequestManager.java:106)
	at com.grubbrrsystems.kiosk.general.utils.GlideRequests.<init>(GlideRequests.java:32)
	at com.bumptech.glide.GeneratedRequestManagerFactory.build(GeneratedRequestManagerFactory.java:18)
	at com.bumptech.glide.manager.RequestManagerRetriever.supportFragmentGet(RequestManagerRetriever.java:438)
	at com.bumptech.glide.manager.RequestManagerRetriever.get(RequestManagerRetriever.java:130)
	at com.bumptech.glide.Glide.with(Glide.java:801)
	at com.grubbrrsystems.kiosk.general.service.MyBaseActivity.DisplayBranchLogo(MyBaseActivity.java:294)
	at com.grubbrrsystems.kiosk.general.NormalItemDetailsActivity.setDetails(NormalItemDetailsActivity.java:195)
	at com.grubbrrsystems.kiosk.general.NormalItemDetailsActivity.onCreate(NormalItemDetailsActivity.java:166)
	at android.app.Activity.performCreate(Activity.java:7878)
	at android.app.Activity.performCreate(Activity.java:7867)
	at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1307)
	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3255)
	... 11 more
Caused by: android.os.RemoteException: Remote stack trace:
	at com.android.server.am.ActivityManagerService.registerReceiver(ActivityManagerService.java:14695)
	at android.app.IActivityManager$Stub.onTransact(IActivityManager.java:1906)
	at com.android.server.am.ActivityManagerService.onTransact(ActivityManagerService.java:2846)
	at android.os.Binder.execTransactInternal(Binder.java:1021)
	at android.os.Binder.execTransact(Binder.java:994)

```


Is there any device specific issue? Please help us to solve this issue.

--- a/library/src/main/java/com/bumptech/glide/manager/DefaultConnectivityMonitor.java
+++ b/library/src/main/java/com/bumptech/glide/manager/DefaultConnectivityMonitor.java
@@ -1,47 +1,20 @@
 package com.bumptech.glide.manager;
 
-import android.annotation.SuppressLint;
-import android.content.BroadcastReceiver;
 import android.content.Context;
-import android.content.Intent;
-import android.content.IntentFilter;
-import android.net.ConnectivityManager;
-import android.net.NetworkInfo;
-import android.util.Log;
 import androidx.annotation.NonNull;
-import com.bumptech.glide.util.Preconditions;
 import com.bumptech.glide.util.Synthetic;
 
-/** Uses {@link android.net.ConnectivityManager} to identify connectivity changes. */
+/**
+ * An Android Lifecycle wrapper that uses {@link SingletonConnectivityReceiver} to observer
+ * connectivity changes, allowing for registration to be removed when our listener is being
+ * destroyed as part of the Android lifecycle.
+ */
 final class DefaultConnectivityMonitor implements ConnectivityMonitor {
-  private static final String TAG = "ConnectivityMonitor";
   private final Context context;
 
   @SuppressWarnings("WeakerAccess")
   @Synthetic
   final ConnectivityListener listener;
-
-  @SuppressWarnings("WeakerAccess")
-  @Synthetic
-  boolean isConnected;
-
-  private boolean isRegistered;
-
-  private final BroadcastReceiver connectivityReceiver =
-      new BroadcastReceiver() {
-        @Override
-        public void onReceive(@NonNull Context context, Intent intent) {
-          boolean wasConnected = isConnected;
-          isConnected = isConnected(context);
-          if (wasConnected != isConnected) {
-            if (Log.isLoggable(TAG, Log.DEBUG)) {
-              Log.d(TAG, "connectivity changed, isConnected: " + isConnected);
-            }
-
-            listener.onConnectivityChanged(isConnected);
-          }
-        }
-      };
 
   DefaultConnectivityMonitor(@NonNull Context context, @NonNull ConnectivityListener listener) {
     this.context = context.getApplicationContext();
@@ -49,56 +22,11 @@ final class DefaultConnectivityMonitor implements ConnectivityMonitor {
   }
 
   private void register() {
-    if (isRegistered) {
-      return;
-    }
-
-    // Initialize isConnected.
-    isConnected = isConnected(context);
-    try {
-      // See #1405
-      context.registerReceiver(
-          connectivityReceiver, new IntentFilter(ConnectivityManager.CONNECTIVITY_ACTION));
-      isRegistered = true;
-    } catch (SecurityException e) {
-      // See #1417, registering the receiver can throw SecurityException.
-      if (Log.isLoggable(TAG, Log.WARN)) {
-        Log.w(TAG, "Failed to register", e);
-      }
-    }
+    SingletonConnectivityReceiver.get(context).register(listener);
   }
 
   private void unregister() {
-    if (!isRegistered) {
-      return;
-    }
-
-    context.unregisterReceiver(connectivityReceiver);
-    isRegistered = false;
-  }
-
-  @SuppressWarnings("WeakerAccess")
-  @Synthetic
-  // Permissions are checked in the factory instead.
-  @SuppressLint("MissingPermission")
-  boolean isConnected(@NonNull Context context) {
-    ConnectivityManager connectivityManager =
-        Preconditions.checkNotNull(
-            (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE));
-    NetworkInfo networkInfo;
-    try {
-      networkInfo = connectivityManager.getActiveNetworkInfo();
-    } catch (RuntimeException e) {
-      // #1405 shows that this throws a SecurityException.
-      // b/70869360 shows that this throws NullPointerException on APIs 22, 23, and 24.
-      // b/70869360 also shows that this throws RuntimeException on API 24 and 25.
-      if (Log.isLoggable(TAG, Log.WARN)) {
-        Log.w(TAG, "Failed to determine connectivity status when connectivity changed", e);
-      }
-      // Default to true;
-      return true;
-    }
-    return networkInfo != null && networkInfo.isConnected();
+    SingletonConnectivityReceiver.get(context).unregister(listener);
   }
 
   @Override

--- a/library/src/main/java/com/bumptech/glide/manager/SingletonConnectivityReceiver.java
+++ b/library/src/main/java/com/bumptech/glide/manager/SingletonConnectivityReceiver.java
@@ -1,0 +1,156 @@
+package com.bumptech.glide.manager;
+
+import android.annotation.SuppressLint;
+import android.content.BroadcastReceiver;
+import android.content.Context;
+import android.content.Intent;
+import android.content.IntentFilter;
+import android.net.ConnectivityManager;
+import android.net.NetworkInfo;
+import android.util.Log;
+import androidx.annotation.GuardedBy;
+import androidx.annotation.NonNull;
+import androidx.annotation.VisibleForTesting;
+import com.bumptech.glide.manager.ConnectivityMonitor.ConnectivityListener;
+import com.bumptech.glide.util.Preconditions;
+import com.bumptech.glide.util.Synthetic;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+/** Uses {@link android.net.ConnectivityManager} to identify connectivity changes. */
+final class SingletonConnectivityReceiver {
+  private static volatile SingletonConnectivityReceiver instance;
+  @Synthetic static final String TAG = "ConnectivityMonitor";
+  // Only accessed on the main thread.
+  @Synthetic boolean isConnected;
+
+  private final BroadcastReceiver connectivityReceiver =
+      new BroadcastReceiver() {
+        @Override
+        public void onReceive(@NonNull Context context, Intent intent) {
+          List<ConnectivityListener> listenersToNotify = null;
+          boolean wasConnected = isConnected;
+          isConnected = isConnected(context);
+          if (wasConnected != isConnected) {
+            if (Log.isLoggable(TAG, Log.DEBUG)) {
+              Log.d(TAG, "connectivity changed, isConnected: " + isConnected);
+            }
+
+            synchronized (SingletonConnectivityReceiver.this) {
+              listenersToNotify = new ArrayList<>(listeners);
+            }
+          }
+          // Make sure that we do not hold our lock while calling our listener. Otherwise we could
+          // deadlock where our listener acquires its lock, then tries to acquire ours elsewhere and
+          // then here we acquire our lock and try to acquire theirs.
+          // The consequence of this is that we may notify a listener after it has been
+          // unregistered in a few specific (unlikely) scenarios. That appears to be safe and is
+          // documented in the unregister method.
+          if (listenersToNotify != null) {
+            for (ConnectivityListener listener : listenersToNotify) {
+              listener.onConnectivityChanged(isConnected);
+            }
+          }
+        }
+      };
+
+  private final Context context;
+
+  @GuardedBy("this")
+  @Synthetic
+  final Set<ConnectivityListener> listeners = new HashSet<ConnectivityListener>();
+
+  @GuardedBy("this")
+  private boolean isRegistered;
+
+  static SingletonConnectivityReceiver get(@NonNull Context context) {
+    if (instance == null) {
+      synchronized (SingletonConnectivityReceiver.class) {
+        if (instance == null) {
+          instance = new SingletonConnectivityReceiver(context);
+        }
+      }
+    }
+    return instance;
+  }
+
+  @VisibleForTesting
+  static void reset() {
+    instance = null;
+  }
+
+  private SingletonConnectivityReceiver(@NonNull Context context) {
+    this.context = context.getApplicationContext();
+  }
+
+  synchronized void register(ConnectivityListener listener) {
+    listeners.add(listener);
+    maybeRegisterReceiver();
+  }
+
+  /**
+   * To avoid holding a lock while notifying listeners, the unregisterd listener may still be
+   * notified about a connectivity change after this method completes if this method is called on a
+   * thread other than the main thread and if a connectivity broadcast is racing with this method.
+   * Callers must handle this case.
+   */
+  synchronized void unregister(ConnectivityListener listener) {
+    listeners.remove(listener);
+    maybeUnregisterReceiver();
+  }
+
+  @GuardedBy("this")
+  private void maybeRegisterReceiver() {
+    if (isRegistered || listeners.isEmpty()) {
+      return;
+    }
+    isConnected = isConnected(context);
+    try {
+      // See #1405
+      context.registerReceiver(
+          connectivityReceiver, new IntentFilter(ConnectivityManager.CONNECTIVITY_ACTION));
+      isRegistered = true;
+    } catch (SecurityException e) {
+      // See #1417, registering the receiver can throw SecurityException.
+      if (Log.isLoggable(TAG, Log.WARN)) {
+        Log.w(TAG, "Failed to register", e);
+      }
+    }
+  }
+
+  @GuardedBy("this")
+  private void maybeUnregisterReceiver() {
+    if (!isRegistered || !listeners.isEmpty()) {
+      return;
+    }
+
+    context.unregisterReceiver(connectivityReceiver);
+    isRegistered = false;
+  }
+
+  @SuppressWarnings("WeakerAccess")
+  @Synthetic
+  // Permissions are checked in the factory instead.
+  @SuppressLint("MissingPermission")
+  boolean isConnected(@NonNull Context context) {
+    ConnectivityManager connectivityManager =
+        Preconditions.checkNotNull(
+            (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE));
+    NetworkInfo networkInfo;
+    try {
+      networkInfo = connectivityManager.getActiveNetworkInfo();
+    } catch (RuntimeException e) {
+      // #1405 shows that this throws a SecurityException.
+      // b/70869360 shows that this throws NullPointerException on APIs 22, 23, and 24.
+      // b/70869360 also shows that this throws RuntimeException on API 24 and 25.
+      if (Log.isLoggable(TAG, Log.WARN)) {
+        Log.w(TAG, "Failed to determine connectivity status when connectivity changed", e);
+      }
+      // Default to true;
+      return true;
+    }
+    return networkInfo != null && networkInfo.isConnected();
+  }
+}

--- a/library/test/src/test/java/com/bumptech/glide/manager/DefaultConnectivityMonitorTest.java
+++ b/library/test/src/test/java/com/bumptech/glide/manager/DefaultConnectivityMonitorTest.java
@@ -17,6 +17,7 @@ import android.net.NetworkInfo;
 import androidx.test.core.app.ApplicationProvider;
 import com.bumptech.glide.manager.DefaultConnectivityMonitorTest.PermissionConnectivityManager;
 import java.util.List;
+import org.junit.After;
 import org.junit.Before;
 import org.junit.Test;
 import org.junit.runner.RunWith;
@@ -44,6 +45,11 @@ public class DefaultConnectivityMonitorTest {
     MockitoAnnotations.initMocks(this);
     monitor = new DefaultConnectivityMonitor(ApplicationProvider.getApplicationContext(), listener);
     harness = new ConnectivityHarness();
+  }
+
+  @After
+  public void tearDown() {
+    SingletonConnectivityReceiver.reset();
   }
 
   @Test
@@ -106,7 +112,7 @@ public class DefaultConnectivityMonitorTest {
     harness.connect();
     harness.broadcast();
 
-    verify(listener).onConnectivityChanged(eq(true));
+    verify(listener).onConnectivityChanged(true);
   }
 
   @Test

